opt: panic in HasColType

SHA: https://github.com/cockroachdb/cockroach/commits/3b0662a83e1226dd4df2e82badacfcfd7440179f

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=sqlsmith/setup=rand-tables/setting=default PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1575698&tab=artifacts#/sqlsmith/setup=rand-tables/setting=default

```
The test failed on branch=release-19.1, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20191106-1575698/sqlsmith/setup=rand-tables/setting=default/run_1
	sqlsmith.go:118,sqlsmith.go:129,test_runner.go:697: ping: dial tcp 35.223.42.55:26257: connect: connection refused
		previous sql:
		INSERT
		INTO
			defaultdb.public.table5 AS tab_3967
		WITH
			with_963 (col_5162)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(B'01011011100111010011010010010111111001001'),
								(B'10111010111011010111010111001000111010110'),
								(B'10100101110011101100101100111101111011101'),
								(B'01010100010101111011100011110011010001100')
						)
							AS tab_3968 (col_5162)
				),
			with_964 (col_5163, col_5164)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(NULL, COALESCE(B'101110000001110000000', B'110101010111000110011')),
								(ltrim(e']\x0e':::STRING::STRING)::STRING, B'011011001010100110011'),
								(e'\U00002603':::STRING, NULL)
						)
							AS tab_3969 (col_5163, col_5164)
				),
			with_965 (col_5165)
				AS (
					SELECT * FROM (VALUES (NULL)) AS tab_3970 (col_5165)
					EXCEPT ALL SELECT * FROM (VALUES ((NULL,))) AS tab_3971 (col_5166)
				)
		SELECT
			tab_3972.col1 AS col_5167,
			tab_3972.col1 AS col_5168,
			'}Ez{':::STRING AS col_5169,
			tab_3972.col1 AS col_5170,
			true AS col_5171
		FROM
			with_963, defaultdb.public.tab_1714 AS tab_3972
		LIMIT
			71:::INT8;

```

Stack trace here:

```
E191106 07:18:20.729303 688 util/log/crash_reporting.go:216  [n1,client=104.196.61.98:42310,user=root] a panic has occurred!
panic: windowNode can't be run in local mode
	panic: windowNode can't be run in local mode [recovered]
	panic: panic while executing 1 statements: INSERT INTO _._._ AS _(_, _, _, _, _) WITH _ (_) AS (SELECT * FROM (VALUES (_:::STRING), (__more5__)) AS _ (_)) , _ (_) AS (SELECT * FROM (VALUES (_)) AS _ (_)) SELECT min(_._::OID)::OID AS _, _:::STRING AS _, min(_._::OID) OVER (PARTITION BY _._, _._)::OID AS _, _:::FLOAT8 AS _, char_length(_::BYTES)::INT8 AS _ FROM _, _._._ AS _ WHERE _ GROUP BY _._, _._; caused by windowNode can't be run in local mode

goroutine 688 [running]:
panic(0x3143900, 0xc012f342a0)
	/usr/local/go/src/runtime/panic.go:556 +0x2cb fp=0xc00adbf778 sp=0xc00adbf6e8 pc=0x735f1b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00a361100, 0x3a522a0, 0xc00aa35d40, 0x2dd3ba0, 0x3a0aa40)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:722 +0x36d fp=0xc00adbf838 sp=0xc00adbf778 pc=0x1fd331d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00a361100, 0x3a522a0, 0xc00aa35d40)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:434 +0x61 fp=0xc00adbf870 sp=0xc00adbf838 pc=0x2161b41
runtime.call32(0x0, 0x343cba0, 0xc0000d0490, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:522 +0x3b fp=0xc00adbf8a0 sp=0xc00adbf870 pc=0x764aeb
panic(0x2dd3ba0, 0x3a0aa40)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9 fp=0xc00adbf930 sp=0xc00adbf8a0 pc=0x735e09
github.com/cockroachdb/cockroach/pkg/sql.(*windowNode).Next(0xc010124000, 0x3a52360, 0xc012f05aa0, 0xc00f7d8a00, 0xc00a361450, 0xc00adbf9a8, 0x713edd, 0x2f10800)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/window.go:175 +0x39 fp=0xc00adbf950 sp=0xc00adbf930 pc=0x213b9d9
github.com/cockroachdb/cockroach/pkg/sql.(*insertNode).BatchedNext(0xc009af7680, 0x3a52360, 0xc012f05aa0, 0xc00f7d8a00, 0xc00a361450, 0xc00adbfa00, 0xc00adbfa10, 0x7df1d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:471 +0xcc fp=0xc00adbf9b8 sp=0xc00adbf950 pc=0x207205c
github.com/cockroachdb/cockroach/pkg/sql.(*rowCountNode).startExec(0xc011719780, 0x3a52360, 0xc012f05aa0, 0xc00f7d8a00, 0xc00a361450, 0x71d165, 0xc00bfea418)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_batch.go:173 +0xd0 fp=0xc00adbfa08 sp=0xc00adbf9b8 pc=0x20bbb30
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2(0x32caa43, 0x5, 0x3a54020, 0xc011719780, 0xc00adbfab8, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:496 +0x55 fp=0xc00adbfa50 sp=0xc00adbfa08 pc=0x216fe95
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1(0xc0108f6de0, 0x32caa43, 0x5, 0x3a54020, 0xc011719780)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:146 +0x5d fp=0xc00adbfa90 sp=0xc00adbfa50 pc=0x217edbd
runtime.call64(0x0, 0x343cc88, 0xc000a7d9b0, 0x2800000028)
	/usr/local/go/src/runtime/asm_amd64.s:523 +0x3b fp=0xc00adbfae0 sp=0xc00adbfa90 pc=0x764b6b
panic(0x2dd3ba0, 0x3a0aa30)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9 fp=0xc00adbfb70 sp=0xc00adbfae0 pc=0x735e09
github.com/cockroachdb/cockroach/pkg/sql.(*windowNode).startExec(0xc010124000, 0x3a52360, 0xc012f05aa0, 0xc00f7d8a00, 0xc00a361450, 0xc010126480, 0xc00adbfc08)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/window.go:171 +0x39 fp=0xc00adbfb90 sp=0xc00adbfb70 pc=0x213b989
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2(0x32cd7c1, 0x6, 0x3a546a0, 0xc010124000, 0xc00f57d500, 0xc00adbfc18)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:496 +0x55 fp=0xc00adbfbd8 sp=0xc00adbfb90 pc=0x216fe95
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1(0xc0108f6de0, 0x32cd7c1, 0x6, 0x3a546a0, 0xc010124000)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:146 +0x5d fp=0xc00adbfc18 sp=0xc00adbfbd8 pc=0x217edbd
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc0108f6de0, 0x3a546a0, 0xc010124000, 0x32cd7c1, 0x6)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:613 +0x1a2 fp=0xc00adc0670 sp=0xc00adbfc18 pc=0x2135572
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc0108f6de0, 0x3a546a0, 0xc010124000, 0x32ccbfd, 0x6)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:113 +0x8d fp=0xc00adc06e0 sp=0xc00adc0670 pc=0x21351fd
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc0108f6de0, 0x3a53c60, 0xc009af7680, 0x32ccbfd, 0x6)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:452 +0x47c6 fp=0xc00adc1138 sp=0xc00adc06e0 pc=0x2139b96
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitConcrete(0xc0108f6de0, 0x3a53c60, 0xc009af7680)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:126 +0x81 fp=0xc00adc1170 sp=0xc00adc1138 pc=0x21353b1
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc0108f6de0, 0x3a54020, 0xc011719780, 0x32caa43, 0x5)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:533 +0x4ceb fp=0xc00adc1bc8 sp=0xc00adc1170 pc=0x213a0bb
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc0108f6de0, 0x3a54020, 0xc011719780, 0x203004, 0x717d89)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:113 +0x8d fp=0xc00adc1c38 sp=0xc00adc1bc8 pc=0x21351fd
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(0x3a52360, 0xc012f05aa0, 0x3a54020, 0xc011719780, 0x0, 0x343d430, 0x0, 0x0, 0x0, 0xc012f05c50, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:77 +0x1a4 fp=0xc00adc1d60 sp=0xc00adc1c38 pc=0x2135104
github.com/cockroachdb/cockroach/pkg/sql.startExec(0x3a52360, 0xc012f05aa0, 0xc00f7d8a00, 0xc00a361450, 0x3a54020, 0xc011719780, 0xc0118540f8, 0xc0118540f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:499 +0x116 fp=0xc00adc1e10 sp=0xc00adc1d60 pc=0x20b52d6
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc00b217400, 0x3a52360, 0xc012f05aa0, 0x5bc4480, 0x2dd2e20)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:124 +0xd9 fp=0xc00adc1e60 sp=0xc00adc1e10 pc=0x20be209
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc00b217400, 0x3a52360, 0xc012f05aa0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:800 +0x52 fp=0xc00adc1ea0 sp=0xc00adc1e60 pc=0x1dfea72
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc01096d0e0, 0x3a52360, 0xc012f05aa0, 0x343cb20, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:626 +0x1e9 fp=0xc00adc1f00 sp=0xc00adc1ea0 pc=0x1dddee9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc00004e8c0, 0xc0108f6840, 0xc00e98bcb0, 0xc00adc2838, 0xc00ab4f080, 0xc00a361510, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:251 +0x8a0 fp=0xc00adc2728 sp=0xc00adc1f00 pc=0x203bb30
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc00004e8c0, 0x3a52360, 0xc012e8fd70, 0xc00a361510, 0xc0108f6840, 0xc00e98bcb0, 0x3a54020, 0xc011719780, 0xc00ab4f080)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:838 +0x227 fp=0xc00adc29e8 sp=0xc00adc2728 pc=0x203f1b7
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc00a361100, 0x3a52360, 0xc012e8fd70, 0xc00a361450, 0x2, 0x7f58d0cb6d28, 0xc011748dc0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1125 +0x283 fp=0xc00adc2a78 sp=0xc00adc29e8 pc=0x1fe5633
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc00a361100, 0x3a52360, 0xc012e8fd70, 0xc00a361450, 0x7f58d0cb6d28, 0xc011748dc0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:961 +0x658 fp=0xc00adc2b78 sp=0xc00adc2a78 pc=0x1fe4508
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc00a361100, 0x3a52360, 0xc012e8fd70, 0x3a579a0, 0xc012a85bd0, 0xc00c523a32, 0x369, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:456 +0xdd0 fp=0xc00adc3358 sp=0xc00adc2b78 pc=0x1fe05f0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc00a361100, 0x3a52360, 0xc012e8fd70, 0x3a579a0, 0xc012a85bd0, 0xc00c523a32, 0x369, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:102 +0x610 fp=0xc00adc3600 sp=0xc00adc3358 pc=0x1fdf3a0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc00a361100, 0x3a522a0, 0xc00aa35d40, 0xc000ba2c38, 0x5400, 0x15000, 0xc000ba2cd0, 0xc00a99d0f0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1186 +0x21d4 fp=0xc00adc3e68 sp=0xc00adc3600 pc=0x1fd6ad4
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000910b00, 0x3a522a0, 0xc00aa35d40, 0xc00a361100, 0x5400, 0x15000, 0xc000ba2cd0, 0xc00a99d0f0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:436 +0xce fp=0xc00adc3ec8 sp=0xc00adc3e68 pc=0x1fd169e
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc00aaf0836, 0xc00aae2680, 0x3a522a0, 0xc00aa35d40, 0xc00a99d0f0, 0xc000910b00, 0xc009e2b200, 0x3a55fe0, 0xc00aae2660, 0xc00aa3d5c0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:580 +0x21f fp=0xc00adc3f68 sp=0xc00adc3ec8 pc=0x225d47f
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1333 +0x1 fp=0xc00adc3f70 sp=0xc00adc3f68 pc=0x766801
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:515 +0x17b
```

Looks like an error at https://github.com/cockroachdb/cockroach/blob/3b0662a83e1226dd4df2e82badacfcfd7440179f/pkg/sql/conn_executor.go#L434

This crashed the gateway node. My guess from the line above is the conn executor isn't setup (or the ConnectionHandler? dunno). But @andreimatei this looks like it might be your cup filled with tea.

--- a/pkg/sql/opt/optbuilder/testdata/union
+++ b/pkg/sql/opt/optbuilder/testdata/union
@@ -1017,3 +1017,11 @@ except
       │                   └── const: 1 [type=int]
       └── tuple [type=tuple{int}]
            └── const: 1 [type=int]
+
+# Regression test for #42216.
+norm
+  SELECT * FROM (VALUES (NULL)) AS t1(a)
+EXCEPT ALL
+  SELECT * FROM (VALUES ((NULL,))) AS t2(b)
+----
+error (42P16): value type unknown cannot be used for table columns

--- a/pkg/sql/opt/optbuilder/union.go
+++ b/pkg/sql/opt/optbuilder/union.go
@@ -166,7 +166,10 @@ func (b *Builder) propagateTypes(dst, src *scope) *scope {
 		srcType := src.cols[i].typ
 		if dstType == types.Unknown && srcType != types.Unknown {
 			// Create a new column which casts the old column to the correct type.
-			colType, _ := coltypes.DatumTypeToColumnType(srcType)
+			colType, err := coltypes.DatumTypeToColumnType(srcType)
+			if err != nil {
+				panic(builderError{err})
+			}
 			castExpr := b.factory.ConstructCast(b.factory.ConstructVariable(dstCols[i].id), colType)
 			b.synthesizeColumn(dst, string(dstCols[i].name), srcType, nil /* expr */, castExpr)
 		} else {

